framework.cpp

```diff
@@ -0,0 +1,403 @@
+#ifndef _METRIC_FRAMEWORK_CPP
+#define _METRIC_FRAMEWORK_CPP
+
+#include <cmath>
+#include <vector>
+//#include <map>
+
+#include "3rdparty/blaze/Blaze.h"
+
+
+namespace metric {
+
+
+
+
+// --------
+// common type trait metafunctions
+
+
+
+template <typename>
+struct determine_container_type  // checks whether container is STL container (1) or Blaze vector (2)
+{
+    constexpr static int code = 0;
+};
+
+template <template <typename, typename> class Container, typename ValueType, typename Allocator>
+struct determine_container_type<Container<ValueType, Allocator>>
+{
+    constexpr static int code = 1; // STL
+};
+
+template <template <typename, bool> class Container, typename ValueType, bool F>
+struct determine_container_type<Container<ValueType, F>>
+{
+    constexpr static int code = 2; // blaze::DynamicVector
+};
+
+template <template <typename, int, int, int, int, int> class Container, typename ValueType, int _Rows, int _Cols, int _Options, int _MaxRows, int _MaxCols>
+struct determine_container_type<Container<ValueType, _Rows, _Cols, _Options, _MaxRows, _MaxCols>>
+{
+    constexpr static int code = 3; // Eigen::Array
+};
+
+
+template<typename C, int = determine_container_type<C>::code>
+struct determine_val_t  // determines type of element both for STL containers and Blaze vectors
+{
+    using type = void;
+};
+
+template<typename C>
+struct determine_val_t<C, 1>
+{
+    using type = typename C::value_type;
+};
+
+template<typename C>
+struct determine_val_t<C, 2>
+{
+    using type = typename C::ElementType;
+};
+
+template<typename C>
+struct determine_val_t<C, 3>
+{
+    using type = typename C::Scalar;
+};
+
+
+
+// --------
+// base classes
+
+
+template <typename rec_t>
+class MetricBase { // all metrics to be derived from
+
+public:
+
+    using Rec_T = rec_t;
+    using Val_T = determine_val_t<Rec_T>;
+
+    MetricBase<rec_t>() {}
+
+    double operator () (const rec_t & a, const rec_t & b) { // to be overridden in derived metrics
```

Comment:
do we agreed to use use single type for all metrics - double?

Reply:
we have the value type of the record and we have some inner type of the algorithm and also a return type. in the past we tried to strictly forward the types: value type --> inner types --> return type.
But now we have a problem, what is the return type of the Euclidean distance of two integer vectors.

Reply:
we could use `std::invoke_result` traits to deduct return type of metric from its inputs, like this
```
using result_type = std::invoke_result<Metric, const RecType&, const RecType&>::type;
```

Reply:
so metric can return arbitrary type, and this does not affect type type of metric, ok. But in this scenario algorithm has to be able to support arbitrary type obtained from metric. In some cases it can be useful: for instance, int instead of double may signal to algorithm that specific group of metrics is used and special processing is needed. Cant observe all consequences of this design yet.. I'll try

Reply:
I've tried the example where second template parameter defaulted to double is added to metric. It sets the return type